# **KLASIFIKASI KELAS JAMUR DAPAT DIMAKAN atau BERACUN Berdasarkan metode terbaik dari pycaret**

##**Deskripsi Data**

data yang digunakan di ambil dari website UCI REPOSITORY : https://archive.ics.uci.edu/dataset/73/mushroom .Dataset mushroom digunakan untuk menganalisis faktor-faktor yang membedakan jamur yang dapat dikonsumsi dengan yang beracun berdasarkan berbagai fitur karakteristik.

**Deskripsi Dataset:**
1. **Jumlah Sampel:** Dataset "Mushroom" terdiri dari 8124 baris data, yang mewakili informasi tentang berbagai jenis jamur.

2. **Missing Value:**
   - Fitur "Stalk Root (Akar Batang)" mengandung nilai yang hilang (missing value) yang ditandai dengan tanda "?". Untuk menangani nilai yang hilang pada fitur ini, dilakukan pengisian dengan modus dari data yang memiliki label yang sama dengan label pada missing value tersebut. Pendekatan ini dipilih karena dataset ini bersifat kategorikal.

3. **One-Hot Encoding:**
   - **One-Hot Encoding:** Fitur-fitur kategorikal, seperti "Cap Shape," "Cap Surface," dan lainnya, dapat diubah menggunakan one-hot encoding. Hal ini mengonversi nilai kategorikal menjadi vektor biner dengan 0 dan 1, memfasilitasi pemrosesan data dalam model pembelajaran.
4. **Fitur- Fitur pada dataset (22 fitur dan label)**
  1. **Label (Edible/Poisonous):**
    - **Deskripsi**: Menandakan apakah jamur tersebut dapat dikonsumsi (edible) atau beracun (poisonous).
    - **Kode:**
      - `edible=e`: Jamur dapat dikonsumsi.
      - `poisonous=p`: Jamur beracun.

  **2. Cap Shape (Bentuk Tudung):**
    - **Deskripsi**: Menunjukkan bentuk fisik dari tudung jamur.
    - **Kode:**
      - `bell=b`: Bentuk seperti lonceng.
      - `conical=c`: Bentuk kerucut.
      - `convex=x`: Bentuk cembung.
      - `flat=f`: Bentuk datar.
      - `knobbed=k`: Bentuk dengan tonjolan.
      - `sunken=s`: Bentuk cekung.

  **3. Cap Surface (Permukaan Tudung):**
    - **Deskripsi**: Menjelaskan tekstur permukaan tudung jamur.
    - **Kode:**
      - `fibrous=f`: Permukaan serat.
      - `grooves=g`: Permukaan berlekuk.
      - `scaly=y`: Permukaan bersisik.
      - `smooth=s`: Permukaan halus.

  **4. Cap Color (Warna Tudung):**
    - **Deskripsi**: Menandakan warna dari tudung jamur.
    - **Kode:**
      - `brown=n`: Coklat.
      - `buff=b`: Coklat muda.
      - `cinnamon=c`: Coklat kayu manis.
      - `gray=g`: Abu-abu.
      - `green=r`: Hijau.
      - `pink=p`: Merah muda.
      - `purple=u`: Ungu.
      - `red=e`: Merah.
      - `white=w`: Putih.
      - `yellow=y`: Kuning.

  **5. Bruises (Memar):**
    - **Deskripsi**: Menandakan apakah jamur tersebut akan memar saat disentuh atau tidak.
    - **Kode:**
      - `bruises=t`: Memar.
      - `no=f`: Tidak memar.

  **6. Odor (Bau):**
    - **Deskripsi**: Menunjukkan aroma yang dihasilkan oleh jamur.
    - **Kode:**
      - `almond=a`: Aroma almond.
      - `anise=l`: Aroma adas manis.
      - `creosote=c`: Aroma creosote.
      - `fishy=y`: Aroma ikan.
      - `foul=f`: Aroma busuk.
      - `musty=m`: Aroma berjamur.
      - `none=n`: Tidak ada aroma.
      - `pungent=p`: Aroma tajam.
      - `spicy=s`: Aroma pedas.

  **7. Gill Attachment (Lampiran Gill):**
    - **Deskripsi**: Menunjukkan bagaimana hubungan gill dengan batang jamur.
    - **Kode:**
      - `attached=a`: Melekat.
      - `descending=d`: Menurun.
      - `free=f`: Bebas.
      - `notched=n`: Tidak rata.

  **8. Gill Spacing (Jarak Gill):**
    - **Deskripsi**: Menunjukkan seberapa rapat gill jamur.
    - **Kode:**
      - `close=c`: Rapat.
      - `crowded=w`: Padat.
      - `distant=d`: Jauh.

  **9. Gill Size (Ukuran Gill):**
    - **Deskripsi**: Menunjukkan ukuran gill jamur.
    - **Kode:**
      - `broad=b`: Lebar.
      - `narrow=n`: Sempit.

  **10. Gill Color (Warna Gill):**
    - **Deskripsi**: Menunjukkan warna dari gill jamur.
    - **Kode:**
      - `black=k`: Hitam.
      - `brown=n`: Coklat.
      - `buff=b`: Coklat muda.
      - `chocolate=h`: Coklat tua.
      - `gray=g`: Abu-abu.
      - `green=r`: Hijau.
      - `orange=o`: Oranye.
      - `pink=p`: Merah muda.
      - `purple=u`: Ungu.
      - `red=e`: Merah.
      - `white=w`: Putih.
      - `yellow=y`: Kuning.

  **11. Stalk Shape (Bentuk Batang):**
    - **Deskripsi**: Menunjukkan apakah batang jamur membesar atau menyempit.
    - **Kode:**
      - `enlarging=e`: Membesar.
      - `tapering=t`: Menyempit.

  **12. Stalk Root (Akar Batang):**
    - **Deskripsi**: Menjelaskan jenis akar batang jamur.
    - **Kode:**
      - `bulbous=b`: Berbentuk bulat.
      - `club=c`: Berbentuk klub.
      - `cup=u`: Berbentuk cangkir.
      - `equal=e`: Sama panjang dengan batang.
      - `rhizomorphs=z`: Berbentuk seperti rizom.
      - `rooted=r`: Mempunyai akar.
      - `missing=?`: Tidak diketahui.

  **13. Stalk Surface Above Ring (Permukaan Batang di Atas Cincin):**
    - **Deskripsi**: Menunjukkan tekstur permukaan batang di atas cincin jamur.
    - **Kode:**
      - `fibrous=f`: Permukaan serat.
      - `scaly=y`: Permukaan bersisik.
      - `silky=k`: Permukaan berkilau.
      - `smooth=s`: Permukaan halus.

  **14. Stalk Surface Below Ring (Permukaan Batang di Bawah Cincin):**
    - **Deskripsi**: Menunjukkan tekstur permukaan batang di bawah cincin jamur.
    - **Kode:**
      - `fibrous=f`: Permukaan serat.
      - `scaly=y`: Permukaan bersisik.
      - `silky=k`: Permukaan berkilau.
      - `smooth=s`: Permukaan halus.

  **15. Stalk Color Above Ring (Warna Batang di Atas Cincin):**
    - **Deskripsi**: Menunjukkan warna batang di atas cincin jamur.
    - **Kode:**
      - `brown=n`: Coklat.
      - `buff=b`: Coklat muda.
      - `cinnamon=c`: Coklat kayu manis.
      - `gray=g`: Abu-abu

  .
      - `orange=o`: Oranye.
      - `pink=p`: Merah muda.
      - `red=e`: Merah.
      - `white=w`: Putih.
      - `yellow=y`: Kuning.

  **16. Stalk Color Below Ring (Warna Batang di Bawah Cincin):**
    - **Deskripsi**: Menunjukkan warna batang di bawah cincin jamur.
    - **Kode:**
      - `brown=n`: Coklat.
      - `buff=b`: Coklat muda.
      - `cinnamon=c`: Coklat kayu manis.
      - `gray=g`: Abu-abu.
      - `orange=o`: Oranye.
      - `pink=p`: Merah muda.
      - `red=e`: Merah.
      - `white=w`: Putih.
      - `yellow=y`: Kuning.

  **17. Veil Type (Tipe Tutup):**
    - **Deskripsi**: Menunjukkan tipe tutup (veil) jamur.
    - **Kode:**
      - `partial=p`: Sebagian.
      - `universal=u`: Universal (menutup seluruhnya).

  **18. Veil Color (Warna Tutup):**
    - **Deskripsi**: Menunjukkan warna tutup jamur.
    - **Kode:**
      - `brown=n`: Coklat.
      - `orange=o`: Oranye.
      - `white=w`: Putih.
      - `yellow=y`: Kuning.

  **19. Ring Number (Jumlah Cincin):**
    - **Deskripsi**: Menunjukkan jumlah cincin pada jamur.
    - **Kode:**
      - `none=n`: Tidak ada.
      - `one=o`: Satu.
      - `two=t`: Dua.

  **20. Ring Type (Tipe Cincin):**
    - **Deskripsi**: Menunjukkan jenis cincin pada jamur.
    - **Kode:**
      - `cobwebby=c`: Berbentuk seperti jaring laba-laba.
      - `evanescent=e`: Cepat menghilang.
      - `flaring=f`: Membentuk bentuk seperti terompet.
      - `large=l`: Besar.
      - `none=n`: Tidak ada.
      - `pendant=p`: Membentuk gantungan.
      - `sheathing=s`: Membentuk lapisan.
      - `zone=z`: Berbentuk zona.

  **21. Spore Print Color (Warna Spora):**
    - **Deskripsi**: Menunjukkan warna spora yang dihasilkan oleh jamur.
    - **Kode:**
      - `black=k`: Hitam.
      - `brown=n`: Coklat.
      - `buff=b`: Coklat muda.
      - `chocolate=h`: Coklat tua.
      - `green=r`: Hijau.
      - `orange=o`: Oranye.
      - `purple=u`: Ungu.
      - `white=w`: Putih.
      - `yellow=y`: Kuning.

  **22. Population (Populasi):**
    - **Deskripsi**: Menunjukkan sebaran populasi jamur di suatu tempat.
    - **Kode:**
      - `abundant=a`: Melimpah.
      - `clustered=c`: Berkumpul.
      - `numerous=n`: Banyak.
      - `scattered=s`: Tersebar.
      - `several=v`: Beberapa.
      - `solitary=y`: Tunggal.

  **23. Habitat (Habitat):**
   - **Deskripsi**: Menunjukkan habitat tempat tumbuhnya jamur.
   - **Kode:**
     - `grasses=g`: Di atas rumput.
     - `leaves=l`: Di atas daun.
     - `meadows=m`: Di padang rumput.
     - `paths=p`: Di jalur.
     - `urban=u`: Di perkotaan.
     - `waste=w`: Di tempat pembuangan.
     - `woods=d`: Di hutan.

# **PRE-PROCESSING**

Dataset yang saya ambil tidak dalam format '.csv' namun dalam format '.data', maka dari itu saya akan lakukan perubahan kedalam format csv serta akan melakukan pengisian pada missing value, melakukan onehot encoding pada dataset supaay menjadi data numerik. selain itu, supaya tidak harus menyambungkan code ini ke google drive, maka data yang telah saya download dari uci saya upload ke repository github saya yaitu repositori 'PSD'dan begitupun data yang telah saya transformasi saya upload di github.

## **Import Library**

disini akan menggunakan beberapa library, namun yang pertama kita memerlukan library pandas untuk menampilkan data dalam dataframe

In [13]:
import pandas as pd #

1. 'import pandas':Ini adalah pernyataan dasar untuk mengimpor library Pandas ke dalam program Python. Dengan menggunakan import pandas, kita membuat seluruh fungsionalitas Pandas dapat diakses di dalam program kita.
2. 'as pd': Klausa ini memberikan alias (pseudonim) kepada library yang diimpor. Dalam hal ini, kita memberi alias "pd" kepada Pandas. Penggunaan alias ini memudahkan kita ketika menggunakan fungsi atau objek dari Pandas, karena kita dapat menggunakan "pd" sebagai kata kunci pendek.

In [14]:
# Membaca dataset
data = pd.read_csv('https://raw.githubusercontent.com/LuluatulMaknunah21-048/PSD/main/agaricus-lepiota.data', header=None)
# Menampilkan DataFrame 'data'
data


0  1  2  3  4  5  6  7  8  9   ... 13 14 15 16 17 18 19 20 21 22
0     p  x  s  n  t  p  f  c  n  k  ...  s  w  w  p  w  o  p  k  s  u
1     e  x  s  y  t  a  f  c  b  k  ...  s  w  w  p  w  o  p  n  n  g
2     e  b  s  w  t  l  f  c  b  n  ...  s  w  w  p  w  o  p  n  n  m
3     p  x  y  w  t  p  f  c  n  n  ...  s  w  w  p  w  o  p  k  s  u
4     e  x  s  g  f  n  f  w  b  k  ...  s  w  w  p  w  o  e  n  a  g
...  .. .. .. .. .. .. .. .. .. ..  ... .. .. .. .. .. .. .. .. .. ..
8119  e  k  s  n  f  n  a  c  b  y  ...  s  o  o  p  o  o  p  b  c  l
8120  e  x  s  n  f  n  a  c  b  y  ...  s  o  o  p  n  o  p  b  v  l
8121  e  f  s  n  f  n  a  c  b  n  ...  s  o  o  p  o  o  p  b  c  l
8122  p  k  y  n  f  y  f  c  n  b  ...  k  w  w  p  w  o  e  w  v  l
8123  e  x  s  n  f  n  a  c  b  y  ...  s  o  o  p  o  o  p  o  c  l

[8124 rows x 23 columns]


1. `pd.read_csv('https://raw.githubusercontent.com/LuluatulMaknunah21-048/PSD/main/agaricus-lepiota.data', header=None)`: Menggunakan fungsi `read_csv` dari Pandas untuk membaca data dari URL yang diberikan. Parameter `header=None` menunjukkan bahwa data tidak memiliki baris judul (header). Hasilnya disimpan dalam DataFrame yang dinamakan 'data'.

2. `data`: Baris terakhir menunjukkan DataFrame 'data'. Dalam lingkungan seperti Jupyter Notebook, menampilkan variabel terakhir di sel akan secara otomatis menampilkan nilai DataFrame tersebut.

In [15]:
data.to_csv('jamur.csv',index=False)
#Menyimpan data dalam bentuk csv, tersimpan dalam google drive jika codenya terhubung dengan google drive satu folder dengan code, begitupun jika neg code menggunakan jupiter/python local.

1. data.to_csv('jamur.csv', index=False): Menggunakan metode to_csv dari Pandas untuk menyimpan DataFrame 'data' ke dalam file CSV dengan nama 'jamur.csv'
2. Parameter index=False menunjukkan bahwa indeks DataFrame tidak akan disertakan sebagai kolom dalam file CSV.


In [16]:
jamur=pd.read_csv('https://raw.githubusercontent.com/LuluatulMaknunah21-048/PSD/main/jamur.csv')
jamur

0  1  2  3  4  5  6  7  8  9  ... 13 14 15 16 17 18 19 20 21 22
0     p  x  s  n  t  p  f  c  n  k  ...  s  w  w  p  w  o  p  k  s  u
1     e  x  s  y  t  a  f  c  b  k  ...  s  w  w  p  w  o  p  n  n  g
2     e  b  s  w  t  l  f  c  b  n  ...  s  w  w  p  w  o  p  n  n  m
3     p  x  y  w  t  p  f  c  n  n  ...  s  w  w  p  w  o  p  k  s  u
4     e  x  s  g  f  n  f  w  b  k  ...  s  w  w  p  w  o  e  n  a  g
...  .. .. .. .. .. .. .. .. .. ..  ... .. .. .. .. .. .. .. .. .. ..
8119  e  k  s  n  f  n  a  c  b  y  ...  s  o  o  p  o  o  p  b  c  l
8120  e  x  s  n  f  n  a  c  b  y  ...  s  o  o  p  n  o  p  b  v  l
8121  e  f  s  n  f  n  a  c  b  n  ...  s  o  o  p  o  o  p  b  c  l
8122  p  k  y  n  f  y  f  c  n  b  ...  k  w  w  p  w  o  e  w  v  l
8123  e  x  s  n  f  n  a  c  b  y  ...  s  o  o  p  o  o  p  o  c  l

[8124 rows x 23 columns]

code diatas digunakan untuk membaca dan menampilkan data yang udah diubah dalam bentuk csv

In [17]:
# Mendapatkan nama-nama kolom dari DataFrame 'jamur'
column_names = jamur.columns

# Menampilkan nama-nama kolom
column_names

Index(['0', '1', '2', '3', '4', '5', '6', '7', '8', '9', '10', '11', '12',
       '13', '14', '15', '16', '17', '18', '19', '20', '21', '22'],
      dtype='object')

karna data yang saya ambil tidak langsung dalam format csv atau tidak dalam bentuk tabular dan ketika saya tampilkan datanya nama kolom tidak langsung terbaca maka saya akan merubah nama kolom dari angka menjadi nama kolom yang sesuai dengan deskripsi data pada website UCI repository.

In [18]:
#Mengganti nama kolom menjadi nama kolom sesuai dengan di UCI Repository
jamur.rename(columns={'0': 'Label', '1': 'cap-shape','2':'cap-surface','3':'cap-color','4':'bruises','5':'odor',
                     '6': 'gill-attachment', '7': 'gill-spacing','8':'gill-size','9':'gill-color','10':'stalk-shape',
                     '11': 'stalk-root', '12': 'stalk-surface-above-ring','13':'stalk-surface-below-ring','14':'stalk-color-above-ring','15':'stalk-color-below-ring',
                     '16': 'veil-type', '17': 'veil-color','18':'ring-number','19':'ring-type','20':'spore-print-color',
                     '21': 'population', '22': 'habitat'},inplace=True)
#menampilkan dataset dengan kolom yang sudah di rename
jamur

Label cap-shape cap-surface cap-color bruises odor gill-attachment  \
0        p         x           s         n       t    p               f   
1        e         x           s         y       t    a               f   
2        e         b           s         w       t    l               f   
3        p         x           y         w       t    p               f   
4        e         x           s         g       f    n               f   
...    ...       ...         ...       ...     ...  ...             ...   
8119     e         k           s         n       f    n               a   
8120     e         x           s         n       f    n               a   
8121     e         f           s         n       f    n               a   
8122     p         k           y         n       f    y               f   
8123     e         x           s         n       f    n               a   

     gill-spacing gill-size gill-color  ... stalk-surface-below-ring  \
0               c         n          k  ...                        s   
1               c         b          k  ...                        s   
2               c         b          n  ...                        s   
3               c         n          n  ...                        s   
4               w         b          k  ...                        s   
...           ...       ...        ...  ...                      ...   
8119            c         b          y  ...                        s   
8120            c         b          y  ...                        s   
8121            c         b          n  ...                        s   
8122            c         n          b  ...                        k   
8123            c         b          y  ...                        s   

     stalk-color-above-ring stalk-color-below-ring veil-type veil-color  \
0                         w                      w         p          w   
1                         w                      w         p          w   
2                         w                      w         p          w   
3                         w                      w         p          w   
4                         w                      w         p          w   
...                     ...                    ...       ...        ...   
8119                      o                      o         p          o   
8120                      o                      o         p          n   
8121                      o                      o         p          o   
8122                      w                      w         p          w   
8123                      o                      o         p          o   

     ring-number ring-type spore-print-color population habitat  
0              o         p                 k          s       u  
1              o         p                 n          n       g  
2              o         p                 n          n       m  
3              o         p                 k          s       u  
4              o         e                 n          a       g  
...          ...       ...               ...        ...     ...  
8119           o         p                 b          c       l  
8120           o         p                 b          v       l  
8121           o         p                 b          c       l  
8122           o         e                 w          v       l  
8123           o         p                 o          c       l  

[8124 rows x 23 columns]

## **Mengisi Missing Value**

bagian ini kita sudah memasuki tahap pengisian missing value yang ada pada kolom stalk-root, dan mengisi dengan nilai modus dari setiap label sesuai dengan label yang dimiliki oleh missing value

In [21]:
#melihat jumlah data dan missing value berdasarkan labelnya pada stalk root
# Menghitung jumlah data dalam setiap kelompok dan menyimpannya dalam kolom baru 'jumlah_data'
summary_data = jamur.groupby(['Label', 'stalk-root']).size().reset_index(name='jumlah_data')
#menampilkan jumlah data
summary_data

Label stalk-root  jumlah_data
0     e          ?          720
1     e          b         1920
2     e          c          512
3     e          e          864
4     e          r          192
5     p          ?         1760
6     p          b         1856
7     p          c           44
8     p          e          256

1. jamur.groupby(['Label', 'stalk-root']).size(): Menggunakan metode groupby pada DataFrame 'jamur' untuk mengelompokkan data berdasarkan kolom 'Label' dan 'stalk-root'. size() menghitung jumlah data dalam setiap kelompok.

2. .reset_index(name='jumlah_data'): Setelah pengelompokan, kita menggunakan reset_index untuk mengembalikan indeks grup sebagai kolom, dan memberikan nama baru 'jumlah_data' pada kolom hasil perhitungan.

In [22]:
# Melihat modus dari label e dan p pada fitur stalk-root

# Menghitung modus untuk 'stalk-root' pada kelas 'e' (edible)
modus_e = jamur[jamur['Label'] == 'e']['stalk-root'].mode().iloc[0]

# Menghitung modus untuk 'stalk-root' pada kelas 'p' (poisonous)
modus_p = jamur[jamur['Label'] == 'p']['stalk-root'].mode().iloc[0]

# Menampilkan modus untuk 'stalk-root' pada kedua kelas
print("Modus stalk-root di kelas e (edible) adalah", modus_e)
print("Modus stalk-root di kelas p (poisonous) adalah", modus_p)


Modus stalk-root di kelas e (edible) adalah b
Modus stalk-root di kelas p (poisonous) adalah b


In [ ]:
# Mengganti nilai '?' dengan modus
# Mengganti nilai '?' dengan modus pada kolom 'stalk-root' untuk kelas 'e' (edible)
jamur.loc[jamur['Label'] == 'e', 'stalk-root'] = jamur.loc[jamur['Label'] == 'e', 'stalk-root'].replace('?', moduse)
# Mengganti nilai '?' dengan modus pada kolom 'stalk-root' untuk kelas 'p' (poisonous)
jamur.loc[jamur['Label'] == 'p', 'stalk-root'] = jamur.loc[jamur['Label'] == 'p', 'stalk-root'].replace('?', modusp)

1. jamur.loc[jamur['Label'] == 'e', 'stalk-root'] = jamur.loc[jamur['Label'] == 'e', 'stalk-root'].replace('?', moduse)
   - Kode ini menggunakan metode `.replace` pada kolom 'stalk-root' untuk subset data di mana label ('Label') adalah 'e' (edible). Nilai '?' pada kolom tersebut diganti dengan nilai modus yang sudah dihitung sebelumnya (`moduse`).

2. jamur.loc[jamur['Label'] == 'p', 'stalk-root'] = jamur.loc[jamur['Label'] == 'p', 'stalk-root'].replace('?', modusp)
   - Kode ini melakukan hal yang serupa dengan langkah sebelumnya, tetapi untuk kelas 'p' (poisonous). Nilai '?' pada kolom 'stalk-root' pada subset data kelas 'p' diganti dengan nilai modus kelas 'p' (`modusp`).

In [23]:
#Memastikan apakah pengubahan nilai missing value sudah berhasil atau belum
missing = jamur.groupby(['Label', 'stalk-root']).size().reset_index(name='jumlah_data')
#menampilkan outputnya
missing

Label stalk-root  jumlah_data
0     e          ?          720
1     e          b         1920
2     e          c          512
3     e          e          864
4     e          r          192
5     p          ?         1760
6     p          b         1856
7     p          c           44
8     p          e          256

1. jamur.groupby(['Label', 'stalk-root']).size(): Menggunakan metode groupby pada DataFrame 'jamur' untuk mengelompokkan data berdasarkan kolom 'Label' dan 'stalk-root'. Metode ini diikuti oleh size() untuk menghitung jumlah data dalam setiap kelompok.

2. .reset_index(name='jumlah_data'): Setelah pengelompokan, kita menggunakan reset_index untuk mengembalikan indeks grup sebagai kolom, dan memberikan nama baru 'jumlah_data' pada kolom hasil perhitungan jumlah data.

## **One-hot Encoding**
karna data yang kita miliki adalah data kategorikal maka data akan dilakukan normalisasi dengan one-hot encoding. One-hot encoding adalah teknik representasi data kategorikal dalam bentuk vektor biner, di mana hanya satu bit bernilai 1 sedangkan yang lainnya bernilai 0. Ini digunakan ketika kita memiliki variabel kategorikal dan ingin mengonversinya menjadi bentuk yang dapat digunakan oleh model machine learning yang memerlukan input numerik.

Contoh one-hot encoding:
Jika kita memiliki variabel kategorikal "Warna" dengan nilai {Merah, Hijau, Biru}, maka setelah one-hot encoding, setiap nilai akan direpresentasikan sebagai vektor biner:

Merah: [1, 0, 0]
Hijau: [0, 1, 0]
Biru: [0, 0, 1]

**Melakukan drop pada kolom label karna kolom label tidak akan dilakukan normalisasi one-hot encoding**

In [25]:
#tidak akan melakukan one-hot encoding pada data label
label= ['Label']
jamur_label= jamur[label]
jamur = jamur.drop(columns=label)

**Melakukan one-hot encoding pada fitur serta menambahkan kembali data label**

In [26]:
# Melakukan One-Hot Encoding pada semua kolom
jamur_encoded = pd.get_dummies(jamur)
# Menggabungkan DataFrame hasil One-Hot Encoding dengan label
jamur_final = pd.concat([jamur_label, jamur_encoded], axis=1)
#menampilkan data yang telah di normalisasi
jamur_final

Label  cap-shape_b  cap-shape_c  cap-shape_f  cap-shape_k  cap-shape_s  \
0        p            0            0            0            0            0   
1        e            0            0            0            0            0   
2        e            1            0            0            0            0   
3        p            0            0            0            0            0   
4        e            0            0            0            0            0   
...    ...          ...          ...          ...          ...          ...   
8119     e            0            0            0            1            0   
8120     e            0            0            0            0            0   
8121     e            0            0            1            0            0   
8122     p            0            0            0            1            0   
8123     e            0            0            0            0            0   

      cap-shape_x  cap-surface_f  cap-surface_g  cap-surface_s  ...  \
0               1              0              0              1  ...   
1               1              0              0              1  ...   
2               0              0              0              1  ...   
3               1              0              0              0  ...   
4               1              0              0              1  ...   
...           ...            ...            ...            ...  ...   
8119            0              0              0              1  ...   
8120            1              0              0              1  ...   
8121            0              0              0              1  ...   
8122            0              0              0              0  ...   
8123            1              0              0              1  ...   

      population_s  population_v  population_y  habitat_d  habitat_g  \
0                1             0             0          0          0   
1                0             0             0          0          1   
2                0             0             0          0          0   
3                1             0             0          0          0   
4                0             0             0          0          1   
...            ...           ...           ...        ...        ...   
8119             0             0             0          0          0   
8120             0             1             0          0          0   
8121             0             0             0          0          0   
8122             0             1             0          0          0   
8123             0             0             0          0          0   

      habitat_l  habitat_m  habitat_p  habitat_u  habitat_w  
0             0          0          0          1          0  
1             0          0          0          0          0  
2             0          1          0          0          0  
3             0          0          0          1          0  
4             0          0          0          0          0  
...         ...        ...        ...        ...        ...  
8119          1          0          0          0          0  
8120          1          0          0          0          0  
8121          1          0          0          0          0  
8122          1          0          0          0          0  
8123          1          0          0          0          0  

[8124 rows x 118 columns]

setelah melakukan normalisasi one-hot encoding kolom yang semula 23 kolom (label, 22 fitur) menjadi 118 kolom ( label, 117 fitur).

In [28]:
#Melihat kolom hasil dari one hot encoding
jamur_final.columns.tolist()

['Label',
 'cap-shape_b',
 'cap-shape_c',
 'cap-shape_f',
 'cap-shape_k',
 'cap-shape_s',
 'cap-shape_x',
 'cap-surface_f',
 'cap-surface_g',
 'cap-surface_s',
 'cap-surface_y',
 'cap-color_b',
 'cap-color_c',
 'cap-color_e',
 'cap-color_g',
 'cap-color_n',
 'cap-color_p',
 'cap-color_r',
 'cap-color_u',
 'cap-color_w',
 'cap-color_y',
 'bruises_f',
 'bruises_t',
 'odor_a',
 'odor_c',
 'odor_f',
 'odor_l',
 'odor_m',
 'odor_n',
 'odor_p',
 'odor_s',
 'odor_y',
 'gill-attachment_a',
 'gill-attachment_f',
 'gill-spacing_c',
 'gill-spacing_w',
 'gill-size_b',
 'gill-size_n',
 'gill-color_b',
 'gill-color_e',
 'gill-color_g',
 'gill-color_h',
 'gill-color_k',
 'gill-color_n',
 'gill-color_o',
 'gill-color_p',
 'gill-color_r',
 'gill-color_u',
 'gill-color_w',
 'gill-color_y',
 'stalk-shape_e',
 'stalk-shape_t',
 'stalk-root_?',
 'stalk-root_b',
 'stalk-root_c',
 'stalk-root_e',
 'stalk-root_r',
 'stalk-surface-above-ring_f',
 'stalk-surface-above-ring_k',
 'stalk-surface-above-ring_s',
 's

In [30]:
#Melihat jumlah data berdasarkan label
jumlah= jamur_final['Label'].value_counts()
#menampilkan jumlah data
jumlah

e    4208
p    3916
Name: Label, dtype: int64

In [32]:
jamur_final.to_csv('datanorm.csv',index=False) #menyimpan dataset dalam bentuk csv

dataset yang telah melewati tahapan pre-processing akan disimpan menjadi data bersih dalam format csv yang kemudian akan dilakukan pemodelan dengan library pycaret.

# **PEMODELAN**

Untuk mencari metode yang terbaik kita akan menggunakan library pycaret dan menyimpan model yang terbaik

**Instalasi Pycaret**

In [ ]:
!pip install pycaret

**Penerapan PYCARET**

In [34]:
# Import library PyCaret untuk tugas klasifikasi
from pycaret.classification import *

# Membaca data dari URL dan menyimpannya dalam bentuk dataframe (df)
df = pd.read_csv('https://raw.githubusercontent.com/LuluatulMaknunah21-048/PSD/main/datanorm.csv')

# Mencetak nama kolom dari dataframe untuk identifikasi variabel target yang benar
print(df.columns)

# Memisahkan data menjadi fitur (X) dan target (y)
X = df.drop(columns=['Label'])
y = df['Label']

# Menginisialisasi eksperimen PyCaret dengan menentukan variabel target, menerapkan PCA, dan menentukan komponen PCA
exp = setup(data=df, target='Label', pca=True, pca_method='linear', pca_components=7)

# Membandingkan kinerja berbagai model klasifikasi dan memilih model terbaik
best_model = compare_models()

# Mencetak model terbaik yang telah dipilih
best_model


Index(['Label', 'cap-shape_b', 'cap-shape_c', 'cap-shape_f', 'cap-shape_k',
       'cap-shape_s', 'cap-shape_x', 'cap-surface_f', 'cap-surface_g',
       'cap-surface_s',
       ...
       'population_s', 'population_v', 'population_y', 'habitat_d',
       'habitat_g', 'habitat_l', 'habitat_m', 'habitat_p', 'habitat_u',
       'habitat_w'],
      dtype='object', length=117)


Processing:   0%|          | 0/65 [00:00<?, ?it/s]

ExtraTreesClassifier(bootstrap=False, ccp_alpha=0.0, class_weight=None,
                     criterion='gini', max_depth=None, max_features='sqrt',
                     max_leaf_nodes=None, max_samples=None,
                     min_impurity_decrease=0.0, min_samples_leaf=1,
                     min_samples_split=2, min_weight_fraction_leaf=0.0,
                     n_estimators=100, n_jobs=-1, oob_score=False,
                     random_state=433, verbose=0, warm_start=False)

1. **Import Library:**
 'from pycaret.classification import *'
   Baris ini mengimpor modul-modul yang diperlukan dari PyCaret untuk tugas klasifikasi.

2. **Membaca Data:**
 'df = pd.read_csv('https://raw.githubusercontent.com/LuluatulMaknunah21-048/PSD/main/datanorm.csv')'
   Membaca data dari URL yang diberikan dan menyimpannya dalam bentuk dataframe (`df`).

3. **Print Nama Kolom:**
 'print(df.columns)'
   Baris ini mencetak nama kolom dari dataframe untuk membantu mengidentifikasi variabel target yang benar.

4. **Split Data:**
 ' X = df.drop(columns=['Label'])
   y = df['Label']'
   Memisahkan data menjadi fitur (`X`) dan target (`y`). Fitur `X` adalah dataframe tanpa kolom 'Label', sedangkan `y` adalah kolom 'Label' sebagai variabel target.

5. **Inisialisasi Eksperimen PyCaret:**
   'exp = setup(data=df, target='Label', pca=True, pca_method='linear', pca_components=7)'
   Menginisialisasi eksperimen PyCaret dengan menggunakan fungsi `setup()`. Variabel target ditentukan sebagai 'Label', dan juga diterapkan Principal Component Analysis (PCA) dengan 7 komponen menggunakan metode linear.

6. **Bandingkan Model dan Cari yang Terbaik:**
  'best_model = compare_models()'
   Membandingkan kinerja berbagai model klasifikasi yang sudah tersedia dan memilih model terbaik berdasarkan metrik evaluasi yang digunakan oleh PyCaret.

7. **Cetak Model Terbaik:**
   'best_model'
   Baris ini mencetak model terbaik yang telah dipilih.

**Buat dan Simpan Model**

In [35]:
#create model terbaik
best_model = create_model('et')

# Menyimpan model terbaik
save_model(best_model, 'et_jamur')

Processing:   0%|          | 0/4 [00:00<?, ?it/s]

Transformation Pipeline and Model Successfully Saved


(Pipeline(memory=Memory(location=None),
          steps=[('label_encoding',
                  TransformerWrapperWithInverse(exclude=None, include=None,
                                                transformer=LabelEncoder())),
                 ('numerical_imputer',
                  TransformerWrapper(exclude=None,
                                     include=['cap-shape_b', 'cap-shape_c',
                                              'cap-shape_f', 'cap-shape_k',
                                              'cap-shape_s', 'cap-shape_x',
                                              'cap-surface_f', 'cap-surface_g',
                                              'cap-surface_s', 'cap-sur...
                  ExtraTreesClassifier(bootstrap=False, ccp_alpha=0.0,
                                       class_weight=None, criterion='gini',
                                       max_depth=None, max_features='sqrt',
                                       max_leaf_nodes=None, max_samples=No

1. **Membuat Model:**
   'best_model = create_model('et')'
   Membuat model menggunakan algoritma Extra Trees ('et'), yang sebelumnya telah dipilih sebagai model terbaik.

2. **Menyimpan Model:**
   'save_model(best_model, 'et_jamur')'
   Menyimpan model ke dalam file dengan nama 'et_jamur'. Ini memungkinkan Anda untuk memuat model ini di lain waktu tanpa perlu melatih ulang.

3. **Memuat Model yang Telah Disimpan:**
   'loaded_model = load_model('et_jamur')'
   Memuat model yang telah disimpan sebelumnya dengan nama 'et_jamur'. Ini dapat berguna jika Anda ingin menggunakan model untuk prediksi pada data baru tanpa melatih ulang model.